File: sdk/ts/tests/registration-requirements.test.ts
import { createServer } from "node:http";
import type { AddressInfo } from "node:net";

import { afterEach, beforeEach, describe, expect, it } from "vitest";

import {
  createWellKnownRequirementsReader,
  ExchangeNotPermitted,
  ManifestNotExchange,
} from "../resolvers/index.ts";

const SCHEMA = `{"type":"object","required":["legal_entity"],"properties":{"legal_entity":{"type":"string"}}}`;

/** serve builds a fetch that answers one manifest body and counts calls. */
function serve(body: string): { fetch: typeof fetch; hits: () => number } {
  let hits = 0;
  const fetchFn = (async () => {
    hits++;
    return new Response(body, { status: 200, headers: { "content-type": "application/json" } });
  }) as unknown as typeof fetch;
  return { fetch: fetchFn, hits: () => hits };
}

function manifest(extra: Record<string, unknown> = {}, role: unknown = "ROLE_EXCHANGE"): string {
  return JSON.stringify({ role, endpoint: "https://exchange.test", ...extra });
}

const reader = (body: string, allow?: (d: string) => boolean) => {
  const s = serve(body);
  return {
    r: createWellKnownRequirementsReader({
      fetch: s.fetch,
      scheme: "http",
      ...(allow !== undefined ? { allow } : {}),
    }),
    hits: s.hits,
  };
};

describe("the registration-requirements reader", () => {
  it("reads the digest and compiles the published schema", async () => {
    const digest = `sha256:${"ab".repeat(32)}`;
    const { r } = reader(
      manifest({ terms_digest: digest, account_registration: { data_schema: JSON.parse(SCHEMA) } }),
    );
    const got = await r.resolveRegistrationRequirements("exchange.test");
    expect(got.termsDigest).toBe(digest);
    expect(got.verdict).toBe("accepted");
    // The schema is the one that was served, not merely present.
    expect(got.schema?.validate({}).length).toBeGreaterThan(0);
    expect(got.schema?.validate({ legal_entity: "Acme" })).toEqual([]);
  });

  it("fetches afresh on every read", async () => {
    const { r, hits } = reader(manifest({ terms_digest: `sha256:${"ab".repeat(32)}` }));
    await r.resolveRegistrationRequirements("exchange.test");
    await r.resolveRegistrationRequirements("exchange.test");
    expect(hits()).toBe(2);
  });

  it("treats publishing neither member as a normal answer", async () => {
    const { r } = reader(manifest());
    const got = await r.resolveRegistrationRequirements("exchange.test");
    expect(got.termsDigest).toBeUndefined();
    expect(got.schema).toBeNull();
    expect(got.verdict).toBe("not_published");
  });

  it("reports a refused schema as a verdict, never as a throw", async () => {
    const { r } = reader(
      manifest({
        account_registration: {
          data_schema: { $schema: "https://json-schema.org/draft/2019-09/schema", type: "object" },
        },
      }),
    );
    const got = await r.resolveRegistrationRequirements("exchange.test");
    expect(got.verdict).toBe("wrong_dialect");
    expect(got.schema).toBeNull();
  });

  // The cap is defined over the bytes AS SERVED. A schema padded past the cap on
  // the wire that would minify under it must be refused here exactly as the
  // Exchange refuses it, or a client pre-checks against a schema nobody enforces.
  it("measures the schema over the served bytes, not a re-encoding", async () => {
    const padded = `{"type":"object",${" ".repeat(20_000)}"title":"x"}`;
    const body = `{"role":"ROLE_EXCHANGE","account_registration":{"data_schema":${padded}}}`;
    const { r } = reader(body);
    const got = await r.resolveRegistrationRequirements("exchange.test");
    expect(got.verdict).toBe("too_large");
  });

  it("refuses before dialling", async () => {
    const bad = reader(manifest());
    await expect(bad.r.resolveRegistrationRequirements("exchange.test/path")).rejects.toThrow();
    expect(bad.hits()).toBe(0);

    const blocked = reader(manifest(), () => false);
    await expect(blocked.r.resolveRegistrationRequirements("exchange.test")).rejects.toBeInstanceOf(
      ExchangeNotPermitted,
    );
    expect(blocked.hits()).toBe(0);
  });

  it("refuses a manifest that is not an Exchange", async () => {
    for (const body of [
      manifest({}, "ROLE_BROKER"),
      manifest({}, "ROLE_AGENT"),
      manifest({}, "ROLE_PUBLISHER"),
      // No role at all. The contract makes the field required, so reading silence
      // as assent would leave the check advisory.
      `{"endpoint":"https://exchange.test"}`,
    ]) {
      const { r } = reader(body);
      await expect(r.resolveRegistrationRequirements("exchange.test")).rejects.toBeInstanceOf(
        ManifestNotExchange,
      );
    }
  });

  it("accepts the role as a proto-JSON number", async () => {
    const { r } = reader(manifest({}, 2));
    await expect(r.resolveRegistrationRequirements("exchange.test")).resolves.toBeDefined();
  });
});

// The transport default, dialled for real.
//
// The domain this reader fetches comes off a RegisterRequest, so it is chosen by the
// caller at runtime rather than by the operator at start-up. That is the provenance
// that takes the SSRF-guarded transport — the same rule the Go oracle states in the
// options struct these readers mirror, and the same threat shape as the offer-derived
// legs. A reader that dialled a caller-named host unguarded would turn `register` into
// a blind GET aimed wherever that value pointed, and `isBareDomain` accepts `localhost`
// and `169.254.169.254` because they are well-formed domains.
//
// Real-dial rather than a stubbed fetch, mirroring resolvers-ssrf-gate.test.ts: a stub
// would assert which function was referenced, and what matters is which addresses the
// process will actually connect to.
describe("the reader's default transport", () => {
  // Both guards are env-driven, so the flags are cleared here rather than inherited
  // from whatever shell ran the suite — the same isolation resolvers-ssrf-gate.test.ts
  // uses, and without it a developer with SKIP_SSRF set sees this fail for the wrong
  // reason.
  const TOUCHED = ["SKIP_SSRF", "ALLOW_INSECURE", "HTTP_PROXY", "HTTPS_PROXY"];
  let saved: Record<string, string | undefined>;
  beforeEach(() => {
    saved = {};
    for (const k of TOUCHED) {
      saved[k] = process.env[k];
      delete process.env[k];
    }
  });
  afterEach(() => {
    for (const k of TOUCHED) {
      if (saved[k] === undefined) delete process.env[k];
      else process.env[k] = saved[k];
    }
  });

  it("refuses a loopback origin the injected transport reaches", async () => {
    const server = createServer((_req, res) => {
      res.writeHead(200, { "content-type": "application/json" });
      res.end(manifest());
    });
    await new Promise<void>((resolve) => server.listen(0, "127.0.0.1", resolve));
    const { port } = server.address() as AddressInfo;
    const host = `127.0.0.1:${port}`;
    try {
      // The default: no fetch injected, so the guarded transport, which refuses a
      // reserved address at dial time.
      const guarded = createWellKnownRequirementsReader({ scheme: "http" });
      await expect(guarded.resolveRegistrationRequirements(host)).rejects.toThrow();

      // The same host, same server, through an injected transport — so the refusal
      // above is the guard's verdict and not an unreachable server.
      const injected = createWellKnownRequirementsReader({
        scheme: "http",
        fetch: (async (url: string) => {
          const r = await fetch(url);
          return { status: r.status, text: () => r.text() };
        }) as never,
      });
      const got = await injected.resolveRegistrationRequirements(host);
      expect(got.verdict).toBe("not_published");
    } finally {
      await new Promise<void>((resolve) => server.close(() => resolve()));
    }
  });
});

File: sdk/go/resolvers/registrationrequirements_test.go
package resolvers_test

import (
	"context"
	"encoding/json"
	"errors"
	"net/http"
	"net/http/httptest"
	"testing"

	"github.com/RAMP-Protocol/protocol/sdk/go/helpers"
	"github.com/RAMP-Protocol/protocol/sdk/go/resolvers"
)

// requirementsManifest serves a manifest carrying the members the registration
// reader looks at. A nil digest or schema omits the member entirely, which is how
// "this Exchange publishes none" looks on the wire. It counts requests, because
// the reader's central promise is that every read is a fresh fetch.
func requirementsManifest(role string, digest *string, schema json.RawMessage, hits *int) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, _ *http.Request) {
		if hits != nil {
			*hits++
		}
		doc := map[string]any{"role": role, "endpoint": "https://exchange.test"}
		if digest != nil {
			doc["terms_digest"] = *digest
		}
		if schema != nil {
			doc["account_registration"] = map[string]any{"data_schema": schema}
		}
		_ = json.NewEncoder(w).Encode(doc)
	})
}

// loopbackReader builds a reader against a loopback manifest server. The default
// client is SSRF-guarded and would refuse the address, so the test injects an
// unguarded one exactly as the endpoint-resolver tests do.
func loopbackReader(allow func(string) bool) *resolvers.WellKnownRequirementsReader {
	return resolvers.NewWellKnownRequirementsReader(resolvers.WellKnownOptions{
		Scheme: "http",
		HTTP:   http.DefaultClient,
		Allow:  allow,
	})
}

const acceptedSchema = `{"type":"object","required":["legal_entity"],` +
	`"properties":{"legal_entity":{"type":"string"}}}`

func TestRequirements_readsDigestAndSchema(t *testing.T) {
	digest := "sha256:" + repeat64("ab")
	srv := httptest.NewServer(requirementsManifest(
		"ROLE_EXCHANGE", &digest, json.RawMessage(acceptedSchema), nil))
	defer srv.Close()

	got, err := loopbackReader(nil).ResolveRegistrationRequirements(
		context.Background(), hostOf(t, srv))
	if err != nil {
		t.Fatalf("resolve requirements: %v", err)
	}
	if got.TermsDigest == nil || *got.TermsDigest != digest {
		t.Fatalf("terms digest = %v, want %q", got.TermsDigest, digest)
	}
	if got.Verdict != helpers.SchemaAccepted {
		t.Fatalf("verdict = %v, want accepted", got.Verdict)
	}
	// The schema is not merely present, it is the one that was served: a payload
	// missing the required member must fail against it.
	if fails := got.Schema.Validate(map[string]any{}); len(fails) == 0 {
		t.Fatal("compiled schema accepted a payload missing a required member")
	}
	if fails := got.Schema.Validate(map[string]any{"legal_entity": "Acme"}); len(fails) != 0 {
		t.Fatalf("compiled schema refused a conforming payload: %v", fails)
	}
}

// The rule this reader exists for: a cached endpoint is fine, a cached digest is
// not. Two reads must produce two fetches, and the second must see the new value.
func TestRequirements_everyReadIsAFreshFetch(t *testing.T) {
	digest := "sha256:" + repeat64("ab")
	hits := 0
	srv := httptest.NewServer(requirementsManifest("ROLE_EXCHANGE", &digest, nil, &hits))
	defer srv.Close()
	r := loopbackReader(nil)
	host := hostOf(t, srv)

	if _, err := r.ResolveRegistrationRequirements(context.Background(), host); err != nil {
		t.Fatalf("first read: %v", err)
	}
	digest = "sha256:" + repeat64("cd")
	got, err := r.ResolveRegistrationRequirements(context.Background(), host)
	if err != nil {
		t.Fatalf("second read: %v", err)
	}
	if hits != 2 {
		t.Fatalf("manifest fetches = %d, want 2; the digest was served from a cache", hits)
	}
	if got.TermsDigest == nil || *got.TermsDigest != digest {
		t.Fatalf("second read returned %v, want the revised %q", got.TermsDigest, digest)
	}
}

func TestRequirements_publishingNeitherIsANormalAnswer(t *testing.T) {
	srv := httptest.NewServer(requirementsManifest("ROLE_EXCHANGE", nil, nil, nil))
	defer srv.Close()

	got, err := loopbackReader(nil).ResolveRegistrationRequirements(
		context.Background(), hostOf(t, srv))
	if err != nil {
		t.Fatalf("an Exchange publishing neither member is not an error: %v", err)
	}
	if got.TermsDigest != nil {
		t.Fatalf("terms digest = %q, want nil", *got.TermsDigest)
	}
	if got.Schema != nil || got.Verdict != helpers.SchemaNotPublished {
		t.Fatalf("schema = %v verdict = %v, want nil / not_published", got.Schema, got.Verdict)
	}
}

// A schema this SDK refuses is reported, never raised: the contract requires a
// client that cannot check locally to send anyway and let the Exchange decide.
func TestRequirements_aRefusedSchemaIsAVerdictNotAnError(t *testing.T) {
	wrongDialect := json.RawMessage(
		`{"$schema":"https://json-schema.org/draft/2019-09/schema","type":"object"}`)
	srv := httptest.NewServer(requirementsManifest("ROLE_EXCHANGE", nil, wrongDialect, nil))
	defer srv.Close()

	got, err := loopbackReader(nil).ResolveRegistrationRequirements(
		context.Background(), hostOf(t, srv))
	if err != nil {
		t.Fatalf("a refused schema must not fail the read: %v", err)
	}
	if got.Verdict != helpers.SchemaWrongDialect {
		t.Fatalf("verdict = %v, want wrong_dialect", got.Verdict)
	}
	if got.Schema != nil {
		t.Fatal("a refused schema must not be returned")
	}
	// And the nil schema must behave as "nothing to enforce" rather than as a veto.
	if fails := got.Schema.Validate(map[string]any{}); len(fails) != 0 {
		t.Fatalf("a refused schema became a local veto: %v", fails)
	}
}

func TestRequirements_refusesBeforeDialling(t *testing.T) {
	hits := 0
	srv := httptest.NewServer(requirementsManifest("ROLE_EXCHANGE", nil, nil, &hits))
	defer srv.Close()
	host := hostOf(t, srv)

	t.Run("a value that is not a bare domain", func(t *testing.T) {
		// A path smuggled into the domain would choose WHAT is fetched, because the
		// URL is built by concatenation.
		_, err := loopbackReader(nil).ResolveRegistrationRequirements(
			context.Background(), host+"/evil")
		if !errors.Is(err, helpers.ErrInvalidHost) {
			t.Fatalf("err = %v, want ErrInvalidHost", err)
		}
	})

	t.Run("a domain the deployment excludes", func(t *testing.T) {
		_, err := loopbackReader(func(string) bool { return false }).
			ResolveRegistrationRequirements(context.Background(), host)
		if !errors.Is(err, resolvers.ErrExchangeNotPermitted) {
			t.Fatalf("err = %v, want ErrExchangeNotPermitted", err)
		}
	})

	if hits != 0 {
		t.Fatalf("manifest fetches = %d, want 0; a refusal reached the network", hits)
	}
}

// Registration requirements are an Exchange's to publish. A manifest claiming
// another role is refused rather than read for members it has no business
// carrying — and so is one naming no role at all, since the contract requires the
// field and reading silence as assent would leave the check advisory.
func TestRequirements_refusesAManifestThatIsNotAnExchange(t *testing.T) {
	for _, role := range []string{"ROLE_BROKER", "ROLE_AGENT", "ROLE_PUBLISHER", ""} {
		name := role
		if name == "" {
			name = "no role at all"
		}
		t.Run(name, func(t *testing.T) {
			srv := httptest.NewServer(requirementsManifest(role, nil, nil, nil))
			defer srv.Close()
			_, err := loopbackReader(nil).ResolveRegistrationRequirements(
				context.Background(), hostOf(t, srv))
			if !errors.Is(err, resolvers.ErrManifestNotExchange) {
				t.Fatalf("err = %v, want ErrManifestNotExchange", err)
			}
		})
	}
}

// proto-JSON lets an enum travel as its number, so a manifest that spells the
// role that way is the same manifest.
func TestRequirements_acceptsTheRoleAsANumber(t *testing.T) {
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, _ *http.Request) {
		_, _ = w.Write([]byte(`{"role":2,"endpoint":"https://exchange.test"}`))
	}))
	defer srv.Close()

	if _, err := loopbackReader(nil).ResolveRegistrationRequirements(
		context.Background(), hostOf(t, srv)); err != nil {
		t.Fatalf("numeric role refused: %v", err)
	}
}

// repeat64 builds a 64-character hex body for a sha256 digest from a 2-char unit.
func repeat64(unit string) string {
	out := ""
	for len(out) < 64 {
		out += unit
	}
	return out[:64]
}
